make FillAsync throw on EOF

geoffkizer opened the pull request:
Rearrange a bit of code so that FillAsync consistently throws an IOException on EOF.

Also, remove read-ahead buffering logic from ReadAsync.

@stephentoub @Priya91 @wfurt

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -847,11 +847,6 @@ private Task WriteToStreamAsync(byte[] buffer, int offset, int count, Cancellati
                     ThrowInvalidHttpResponse();
                 }
-
-                if (remaining == _readLength - _readOffset)
```
Is `remaining` still used after removing this, or can you delete it entirely?

geoffkizer replied:
You're right, it's not.  Will delete.

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -877,18 +872,11 @@ private Task WriteToStreamAsync(byte[] buffer, int offset, int count, Cancellati
-                    break;
-                }
-                remaining = newRemaining;
+                remaining = _readLength - _readOffset;
```
The way this is written, `remaining` is going to be lifted to the state machine.  Can we just get rid of it entirely and replace `if (remaining >= 2)` with `if (_readLength - _readOffset >= 2)`, or at the very least move the `int remaining = _readLength - _readOffset;` to inside the while block and remove this line from here?

geoffkizer replied:
Will do

stephentoub commented on src/System.Net.Http/src/System/Net/Http/Managed/HttpConnection.cs:
```diff
@@ -945,6 +933,11 @@ private Task FillAsync(CancellationToken cancellationToken)
                 if (NetEventSource.IsEnabled) Trace($"Received {bytesRead} bytes.");
+                if (bytesRead == 0)
+                {
+                    throw new IOException(SR.net_http_invalid_response);
```
Consider adding a comment about why we throw in this case.  Presumably it's because FillAsync is only ever called when we're expecting more data, and so lack of any data means something went wrong.

geoffkizer commented:
FYI, I noticed I introduced a bug in ReadCrLfAsync: we weren't properly throwing when we didn't see '\r\n'.  Fixed now.

Interesting that tests did not catch this.

stephentoub commented:
>Interesting that tests did not catch this.

We're pretty light on tests for the protocol itself, probably based on the assumption that that's all handled by the underlying HTTP implementation, which this now is :smile:
